// @elizaos/plugin-google-genai/dist/index.js
// src/index.ts
import { logger as logger7, ModelType as ModelType3 } from "@elizaos/core";

// src/init.ts
import { GoogleGenAI as GoogleGenAI2 } from "@google/genai";
import { logger as logger2 } from "@elizaos/core";

// src/utils/config.ts
import { GoogleGenAI, HarmCategory, HarmBlockThreshold } from "@google/genai";
import { logger } from "@elizaos/core";
function getSetting(runtime, key, defaultValue) {
  return runtime.getSetting(key) ?? process.env[key] ?? defaultValue;
}
function getApiKey(runtime) {
  return getSetting(runtime, "GOOGLE_GENERATIVE_AI_API_KEY");
}
function getSmallModel(runtime) {
  return getSetting(runtime, "GOOGLE_SMALL_MODEL") ?? getSetting(runtime, "SMALL_MODEL", "gemini-2.0-flash-001") ?? "gemini-2.0-flash-001";
}
function getLargeModel(runtime) {
  return getSetting(runtime, "GOOGLE_LARGE_MODEL") ?? getSetting(runtime, "LARGE_MODEL", "gemini-2.5-pro-preview-03-25") ?? "gemini-2.5-pro-preview-03-25";
}
function getImageModel(runtime) {
  return getSetting(runtime, "GOOGLE_IMAGE_MODEL") ?? getSetting(runtime, "IMAGE_MODEL", "gemini-2.5-pro-preview-03-25") ?? "gemini-2.5-pro-preview-03-25";
}
function getEmbeddingModel(runtime) {
  return getSetting(runtime, "GOOGLE_EMBEDDING_MODEL", "text-embedding-004") ?? "text-embedding-004";
}
function createGoogleGenAI(runtime) {
  const apiKey = getApiKey(runtime);
  if (!apiKey) {
    logger.error("Google Generative AI API Key is missing");
    return null;
  }
  return new GoogleGenAI({ apiKey });
}
function getSafetySettings() {
  return [
    {
      category: HarmCategory.HARM_CATEGORY_HARASSMENT,
      threshold: HarmBlockThreshold.BLOCK_MEDIUM_AND_ABOVE
    },
    {
      category: HarmCategory.HARM_CATEGORY_HATE_SPEECH,
      threshold: HarmBlockThreshold.BLOCK_MEDIUM_AND_ABOVE
    },
    {
      category: HarmCategory.HARM_CATEGORY_SEXUALLY_EXPLICIT,
      threshold: HarmBlockThreshold.BLOCK_MEDIUM_AND_ABOVE
    },
    {
      category: HarmCategory.HARM_CATEGORY_DANGEROUS_CONTENT,
      threshold: HarmBlockThreshold.BLOCK_MEDIUM_AND_ABOVE
    }
  ];
}

// src/init.ts
function initializeGoogleGenAI(_config, runtime) {
  void (async () => {
    try {
      const apiKey = getApiKey(runtime);
      if (!apiKey) {
        logger2.warn(
          "GOOGLE_GENERATIVE_AI_API_KEY is not set in environment - Google AI functionality will be limited"
        );
        return;
      }
      try {
        const genAI = new GoogleGenAI2({ apiKey });
        const modelList = await genAI.models.list();
        const models = [];
        for await (const model of modelList) {
          models.push(model);
        }
        logger2.log(`Google AI API key validated successfully. Available models: ${models.length}`);
      } catch (fetchError) {
        const message = fetchError instanceof Error ? fetchError.message : String(fetchError);
        logger2.warn(`Error validating Google AI API key: ${message}`);
        logger2.warn("Google AI functionality will be limited until a valid API key is provided");
      }
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      logger2.warn(
        `Google AI plugin configuration issue: ${message} - You need to configure the GOOGLE_GENERATIVE_AI_API_KEY in your environment variables`
      );
    }
  })();
}

// src/models/text.ts
import { logger as logger3, ModelType } from "@elizaos/core";

// src/utils/events.ts
import { EventType } from "@elizaos/core";
function emitModelUsageEvent(runtime, type, prompt, usage) {
  runtime.emitEvent(EventType.MODEL_USED, {
    provider: "google",
    type,
    prompt,
    tokens: {
      prompt: usage.promptTokens,
      completion: usage.completionTokens,
      total: usage.totalTokens
    }
  });
}

// src/utils/tokenization.ts
async function countTokens(text) {
  return Math.ceil(text.length / 4);
}

// src/models/text.ts
async function handleTextSmall(runtime, {
  prompt,
  stopSequences = [],
  maxTokens = 8192,
  temperature = 0.7,
  frequencyPenalty = 0.7,
  presencePenalty = 0.7
}) {
  const genAI = createGoogleGenAI(runtime);
  if (!genAI) {
    throw new Error("Google Generative AI client not initialized");
  }
  const modelName = getSmallModel(runtime);
  logger3.log(`[TEXT_SMALL] Using model: ${modelName}`);
  logger3.debug(`[TEXT_SMALL] Prompt: ${prompt}`);
  try {
    const systemInstruction = runtime.character.system || void 0;
    const response = await genAI.models.generateContent({
      model: modelName,
      contents: prompt,
      config: {
        temperature,
        topK: 40,
        topP: 0.95,
        maxOutputTokens: maxTokens,
        stopSequences,
        safetySettings: getSafetySettings(),
        ...systemInstruction && { systemInstruction }
      }
    });
    const text = response.text || "";
    const promptTokens = await countTokens(prompt);
    const completionTokens = await countTokens(text);
    emitModelUsageEvent(runtime, ModelType.TEXT_SMALL, prompt, {
      promptTokens,
      completionTokens,
      totalTokens: promptTokens + completionTokens
    });
    return text;
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    logger3.error(`[TEXT_SMALL] Error: ${message}`);
    throw error;
  }
}
async function handleTextLarge(runtime, {
  prompt,
  stopSequences = [],
  maxTokens = 8192,
  temperature = 0.7,
  frequencyPenalty = 0.7,
  presencePenalty = 0.7
}) {
  const genAI = createGoogleGenAI(runtime);
  if (!genAI) {
    throw new Error("Google Generative AI client not initialized");
  }
  const modelName = getLargeModel(runtime);
  logger3.log(`[TEXT_LARGE] Using model: ${modelName}`);
  logger3.debug(`[TEXT_LARGE] Prompt: ${prompt}`);
  try {
    const systemInstruction = runtime.character.system || void 0;
    const response = await genAI.models.generateContent({
      model: modelName,
      contents: prompt,
      config: {
        temperature,
        topK: 40,
        topP: 0.95,
        maxOutputTokens: maxTokens,
        stopSequences,
        safetySettings: getSafetySettings(),
        ...systemInstruction && { systemInstruction }
      }
    });
    const text = response.text || "";
    const promptTokens = await countTokens(prompt);
    const completionTokens = await countTokens(text);
    emitModelUsageEvent(runtime, ModelType.TEXT_LARGE, prompt, {
      promptTokens,
      completionTokens,
      totalTokens: promptTokens + completionTokens
    });
    return text;
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    logger3.error(`[TEXT_LARGE] Error: ${message}`);
    throw error;
  }
}

// src/models/embedding.ts
import { logger as logger4, ModelType as ModelType2 } from "@elizaos/core";
async function handleTextEmbedding(runtime, params) {
  const genAI = createGoogleGenAI(runtime);
  if (!genAI) {
    throw new Error("Google Generative AI client not initialized");
  }
  const embeddingModelName = getEmbeddingModel(runtime);
  logger4.debug(`[TEXT_EMBEDDING] Using model: ${embeddingModelName}`);
  if (params === null) {
    logger4.debug("Creating test embedding for initialization");
    const dimension = 768;
    const testVector = Array(dimension).fill(0);
    testVector[0] = 0.1;
    return testVector;
  }
  let text;
  if (typeof params === "string") {
    text = params;
  } else if (typeof params === "object" && params.text) {
    text = params.text;
  } else {
    logger4.warn("Invalid input format for embedding");
    const dimension = 768;
    const fallbackVector = Array(dimension).fill(0);
    fallbackVector[0] = 0.2;
    return fallbackVector;
  }
  if (!text.trim()) {
    logger4.warn("Empty text for embedding");
    const dimension = 768;
    const emptyVector = Array(dimension).fill(0);
    emptyVector[0] = 0.3;
    return emptyVector;
  }
  try {
    const response = await genAI.models.embedContent({
      model: embeddingModelName,
      contents: text
    });
    const embedding = response.embeddings?.[0]?.values || [];
    const promptTokens = await countTokens(text);
    emitModelUsageEvent(runtime, ModelType2.TEXT_EMBEDDING, text, {
      promptTokens,
      completionTokens: 0,
      totalTokens: promptTokens
    });
    logger4.log(`Got embedding with length ${embedding.length}`);
    return embedding;
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    logger4.error(`Error generating embedding: ${message}`);
    const dimension = 768;
    const errorVector = Array(dimension).fill(0);
    errorVector[0] = 0.6;
    return errorVector;
  }
}

// src/models/image.ts
import { logger as logger5 } from "@elizaos/core";
import { fetch } from "undici";
async function handleImageDescription(runtime, params) {
  const genAI = createGoogleGenAI(runtime);
  if (!genAI) {
    throw new Error("Google Generative AI client not initialized");
  }
  let imageUrl;
  let promptText;
  const modelName = getImageModel(runtime);
  logger5.log(`[IMAGE_DESCRIPTION] Using model: ${modelName}`);
  if (typeof params === "string") {
    imageUrl = params;
    promptText = "Please analyze this image and provide a title and detailed description.";
  } else {
    imageUrl = params.imageUrl;
    promptText = params.prompt || "Please analyze this image and provide a title and detailed description.";
  }
  try {
    const imageResponse = await fetch(imageUrl);
    if (!imageResponse.ok) {
      throw new Error(`Failed to fetch image: ${imageResponse.statusText}`);
    }
    const imageData = await imageResponse.arrayBuffer();
    const base64Image = Buffer.from(imageData).toString("base64");
    const contentType = imageResponse.headers.get("content-type") || "image/jpeg";
    const response = await genAI.models.generateContent({
      model: modelName,
      contents: [
        {
          role: "user",
          parts: [
            { text: promptText },
            {
              inlineData: {
                mimeType: contentType,
                data: base64Image
              }
            }
          ]
        }
      ],
      config: {
        temperature: 0.7,
        topK: 40,
        topP: 0.95,
        maxOutputTokens: 8192,
        safetySettings: getSafetySettings()
      }
    });
    const responseText = response.text || "";
    logger5.log("Received response for image description");
    const isCustomPrompt = typeof params === "object" && params.prompt && params.prompt !== "Please analyze this image and provide a title and detailed description.";
    if (isCustomPrompt) {
      return responseText;
    }
    try {
      const jsonResponse = JSON.parse(responseText);
      if (jsonResponse.title && jsonResponse.description) {
        return jsonResponse;
      }
    } catch (e) {
      logger5.debug(`Parsing as JSON failed, processing as text: ${e}`);
    }
    const titleMatch = responseText.match(/title[:\s]+(.+?)(?:\n|$)/i);
    const title = titleMatch?.[1]?.trim() || "Image Analysis";
    const description = responseText.replace(/title[:\s]+(.+?)(?:\n|$)/i, "").trim();
    return { title, description };
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    logger5.error(`Error analyzing image: ${message}`);
    return {
      title: "Failed to analyze image",
      description: `Error: ${message}`
    };
  }
}

// src/models/object.ts
import { logger as logger6 } from "@elizaos/core";
async function generateObjectByModelType(runtime, params, modelType, getModelFn) {
  const genAI = createGoogleGenAI(runtime);
  if (!genAI) {
    throw new Error("Google Generative AI client not initialized");
  }
  const modelName = getModelFn(runtime);
  const temperature = params.temperature ?? 0.1;
  logger6.info(`Using ${modelType} model: ${modelName}`);
  try {
    let enhancedPrompt = params.prompt;
    if (params.schema) {
      enhancedPrompt += `

Please respond with a JSON object that follows this schema:
${JSON.stringify(params.schema, null, 2)}`;
    }
    const response = await genAI.models.generateContent({
      model: modelName,
      contents: enhancedPrompt,
      config: {
        temperature,
        topK: 40,
        topP: 0.95,
        maxOutputTokens: 8192,
        responseMimeType: "application/json",
        safetySettings: getSafetySettings()
      }
    });
    const text = response.text || "";
    const promptTokens = await countTokens(enhancedPrompt);
    const completionTokens = await countTokens(text);
    emitModelUsageEvent(runtime, modelType, params.prompt, {
      promptTokens,
      completionTokens,
      totalTokens: promptTokens + completionTokens
    });
    try {
      const parsedResult = JSON.parse(text);
      return parsedResult;
    } catch (parseError) {
      logger6.error(
        `Failed to parse JSON response: ${parseError instanceof Error ? parseError.message : String(parseError)}`
      );
      const jsonMatch = text.match(/\{[\s\S]*\}/);
      if (jsonMatch) {
        try {
          const extractedResult = JSON.parse(jsonMatch[0]);
          return extractedResult;
        } catch (secondParseError) {
          throw new Error("Failed to parse JSON from response");
        }
      }
      throw parseError;
    }
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    logger6.error(`[generateObject] Error: ${message}`);
    throw error;
  }
}
async function handleObjectSmall(runtime, params) {
  return generateObjectByModelType(runtime, params, "OBJECT_SMALL", getSmallModel);
}
async function handleObjectLarge(runtime, params) {
  return generateObjectByModelType(runtime, params, "OBJECT_LARGE", getLargeModel);
}

// src/index.ts
import { GoogleGenAI as GoogleGenAI3 } from "@google/genai";
var googleGenAIPlugin = {
  name: "google-genai",
  description: "Google Generative AI plugin for Gemini models",
  config: {
    GOOGLE_GENERATIVE_AI_API_KEY: process.env.GOOGLE_GENERATIVE_AI_API_KEY,
    GOOGLE_SMALL_MODEL: process.env.GOOGLE_SMALL_MODEL,
    GOOGLE_LARGE_MODEL: process.env.GOOGLE_LARGE_MODEL,
    GOOGLE_IMAGE_MODEL: process.env.GOOGLE_IMAGE_MODEL,
    GOOGLE_EMBEDDING_MODEL: process.env.GOOGLE_EMBEDDING_MODEL,
    SMALL_MODEL: process.env.SMALL_MODEL,
    LARGE_MODEL: process.env.LARGE_MODEL,
    IMAGE_MODEL: process.env.IMAGE_MODEL
  },
  async init(_config, runtime) {
    initializeGoogleGenAI(_config, runtime);
  },
  models: {
    [ModelType3.TEXT_SMALL]: async (runtime, params) => {
      return handleTextSmall(runtime, params);
    },
    [ModelType3.TEXT_LARGE]: async (runtime, params) => {
      return handleTextLarge(runtime, params);
    },
    [ModelType3.TEXT_EMBEDDING]: async (runtime, params) => {
      return handleTextEmbedding(runtime, params);
    },
    [ModelType3.IMAGE_DESCRIPTION]: async (runtime, params) => {
      return handleImageDescription(runtime, params);
    },
    [ModelType3.OBJECT_SMALL]: async (runtime, params) => {
      return handleObjectSmall(runtime, params);
    },
    [ModelType3.OBJECT_LARGE]: async (runtime, params) => {
      return handleObjectLarge(runtime, params);
    }
  },
  tests: [
    {
      name: "google_genai_plugin_tests",
      tests: [
        {
          name: "google_test_api_key_validation",
          fn: async (runtime) => {
            const apiKey = getApiKey(runtime);
            if (!apiKey) {
              throw new Error("GOOGLE_GENERATIVE_AI_API_KEY not set");
            }
            const genAI = new GoogleGenAI3({ apiKey });
            const modelList = await genAI.models.list();
            const models = [];
            for await (const model of modelList) {
              models.push(model);
            }
            logger7.log(`Available models: ${models.length}`);
          }
        },
        {
          name: "google_test_text_embedding",
          fn: async (runtime) => {
            try {
              const embedding = await runtime.useModel(ModelType3.TEXT_EMBEDDING, {
                text: "Hello, world!"
              });
              logger7.log(`Embedding dimension: ${embedding.length}`);
              if (embedding.length === 0) {
                throw new Error("Failed to generate embedding");
              }
            } catch (error) {
              const message = error instanceof Error ? error.message : String(error);
              logger7.error(`Error in test_text_embedding: ${message}`);
              throw error;
            }
          }
        },
        {
          name: "google_test_text_small",
          fn: async (runtime) => {
            try {
              const text = await runtime.useModel(ModelType3.TEXT_SMALL, {
                prompt: "What is the nature of reality in 10 words?"
              });
              if (text.length === 0) {
                throw new Error("Failed to generate text");
              }
              logger7.log("Generated with TEXT_SMALL:", text);
            } catch (error) {
              const message = error instanceof Error ? error.message : String(error);
              logger7.error(`Error in test_text_small: ${message}`);
              throw error;
            }
          }
        },
        {
          name: "google_test_text_large",
          fn: async (runtime) => {
            try {
              const text = await runtime.useModel(ModelType3.TEXT_LARGE, {
                prompt: "Explain quantum mechanics in simple terms."
              });
              if (text.length === 0) {
                throw new Error("Failed to generate text");
              }
              logger7.log("Generated with TEXT_LARGE:", text.substring(0, 100) + "...");
            } catch (error) {
              const message = error instanceof Error ? error.message : String(error);
              logger7.error(`Error in test_text_large: ${message}`);
              throw error;
            }
          }
        },
        {
          name: "google_test_image_description",
          fn: async (runtime) => {
            try {
              const result = await runtime.useModel(
                ModelType3.IMAGE_DESCRIPTION,
                "https://upload.wikimedia.org/wikipedia/commons/thumb/1/1c/Vitalik_Buterin_TechCrunch_London_2015_%28cropped%29.jpg/537px-Vitalik_Buterin_TechCrunch_London_2015_%28cropped%29.jpg"
              );
              if (result && typeof result === "object" && "title" in result && "description" in result) {
                logger7.log("Image description:", JSON.stringify(result));
              } else {
                logger7.error("Invalid image description result format:", result);
              }
            } catch (error) {
              const message = error instanceof Error ? error.message : String(error);
              logger7.error(`Error in test_image_description: ${message}`);
              throw error;
            }
          }
        },
        {
          name: "google_test_object_generation",
          fn: async (runtime) => {
            try {
              const schema = {
                type: "object",
                properties: {
                  name: { type: "string" },
                  age: { type: "number" },
                  hobbies: { type: "array", items: { type: "string" } }
                },
                required: ["name", "age", "hobbies"]
              };
              const result = await runtime.useModel(ModelType3.OBJECT_SMALL, {
                prompt: "Generate a person profile with name, age, and hobbies.",
                schema
              });
              logger7.log("Generated object:", result);
              if (!result.name || !result.age || !result.hobbies) {
                throw new Error("Generated object missing required fields");
              }
            } catch (error) {
              const message = error instanceof Error ? error.message : String(error);
              logger7.error(`Error in test_object_generation: ${message}`);
              throw error;
            }
          }
        }
      ]
    }
  ]
};
var index_default = googleGenAIPlugin;
export {
  index_default as default,
  googleGenAIPlugin
};
